### PR TITLE
ARROW-9648: [C++] Added compression level parameter to LZ4_FRAME compression codec

### PR DESCRIPTION
**Pull request by joosthooz**
This will only work for the Arrow IPC format, not for Parquet, because Parquet uses the raw LZ4 format instead of the framed format. 

Still to figure out:

- [x]  Do we need to add tests to test if setting the compression level works as intended? (Or is this already covered by tests for the other formats that support levels)

- [x] Figure out the raw format for Parquet files
The raw format does not support the compression level, but supports an acceleration value. This value works in opposite direction; increasing it speeds up the compression in expense of the ratio. The value can be set from 1 to 65537, so that's also completely different from the compression level in the framing format (1 to 12).
Update: as it turns out, the compression level is linked to the compressor variant, not to whether or not framing is used. Now, when using `lz4_raw` (as used by Parquet), the fast compressor is chosen when `compression_level < 3` and the high compression version is chosen when `compression_level >= 3`. The latter supports compression levels so the value is passed on to it. This behavior is similar to what happens in the lz4 CLI program (see for example https://github.com/lz4/lz4/blob/4c9431e9af596af0556e5da0ae99305bafb2b10b/lib/lz4frame.c#L815).

### CODE REVIEW

File: cpp/src/arrow/util/compression_lz4.cc
Comment (bkietz) [inline cpp/src/arrow/util/compression_lz4.cc:56]: if this argument is intentionally ignored, please make an explanatory comment

File: cpp/src/arrow/util/compression_lz4.cc
Comment (bkietz) [inline cpp/src/arrow/util/compression_lz4.cc:52]: Please add a comment explaining this macro (or at least I assume it's a macro and not a constant). If it isn't completely trivial, I'd recommend ensuring that it is only evaluated once:
```suggestion
  static LZ4F_preferences_t prefs = LZ4F_INIT_PREFERENCES;
```

File: cpp/src/arrow/util/compression_lz4.cc
Comment (bkietz) [inline cpp/src/arrow/util/compression_lz4.cc:259]: If there's a possibility that `compression_level` could be overridden, it seems that same override should be applied to `prefs_`
```suggestion
                               prefs_(DefaultPreferences(compression_level_)) {}
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Looks like a good start, thanks for contributing!

Could you add a case or two to `TestCodecMisc.SpecifyCompressionLevel`? That test doesn't currently touch LZ4_FRAME. The other tests seem fairly generic and should pick up the new support for compression level automatically

File: cpp/src/arrow/util/compression_lz4.cc
Comment (joosthooz) [reply cpp/src/arrow/util/compression_lz4.cc]: That function turned out to be not necessary, it is now removed

File: cpp/src/arrow/util/compression_lz4.cc
Comment (joosthooz) [reply cpp/src/arrow/util/compression_lz4.cc]: Actually it seems that using the memset is also fine, so I reverted back to using that

File: cpp/src/arrow/util/compression_lz4.cc
Comment (joosthooz) [reply cpp/src/arrow/util/compression_lz4.cc]: Ok changed, thanks

File: cpp/src/arrow/util/compression_test.cc
Comment (bkietz) [inline cpp/src/arrow/util/compression_test.cc:382]:
```suggestion
      {Compression::LZ4_FRAME, 9, true}, {Compression::LZ4_FRAME, -99, false}};
```

Comment (joosthooz) [reply cpp/src/arrow/util/compression_test.cc]: Fixed, and also removed the testcase with invalid value, because LZ4 will rectify that internally (and not fail)

**Review by bkietz — CHANGES_REQUESTED** [review]
Would it make sense to define LZ4 (parquet's LZ4_RAW)'s `compression_level = 2**16 - acceleration_level`? That should establish the expected tradeoff of greater compression ratio for speed with higher compression_level

File: cpp/src/arrow/util/compression_lz4.cc
Comment (joosthooz) [reply cpp/src/arrow/util/compression_lz4.cc]: Ok changed

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/util/compression_lz4.cc
Comment (pitrou) [inline cpp/src/arrow/util/compression_lz4.cc:45]: The same constant is already defined in the `.h`, why redefine it here?

Comment (joosthooz) [reply cpp/src/arrow/util/compression_lz4.cc:45]: Yes that's a duplicate, I'll remove it. Actually, wouldn't it be nicer to move the min/max values into the `.h` file too? (also see `compression_bz2.h`, it has the min/max values in the `.cc` file)

**Comment by pitrou** [comment]
Hmm, instead of requiring a source build of LZ4 on older distros, what do you think about doing a compile-time switch? LZ4 exposes `LZ4_VERSION_MAJOR`, `LZ4_VERSION_MINOR` and `LZ4_VERSION_RELEASE` C macro definitions.

**Comment by joosthooz** [comment]
@pitrou without building from source it would become difficult to reliably find the max compression level value; the API call for it was introduced in 1.8.0 and before that the name of the `LZ4HC_CLEVEL_MAX` macro was changed without changing the version number.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (bkietz) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2116]: Do we need this now that the call to `LZ4F_compressionLevel_max` is guarded by an `#if`?

Comment (joosthooz) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2116]: No we do not, I'll revert these changes.

File: cpp/src/arrow/util/compression_lz4.cc
Comment (bkietz) [inline cpp/src/arrow/util/compression_lz4.cc:338]: Since we coerce `kUse$x -> k$x`, it seems this should be:
```suggestion
  Lz4Codec() : compression_level_(kDefaultCompressionLevel) {}
```

... but this constructor isn't used anymore AFAICT, so we could just delete it

**Review by bkietz — APPROVED** [review]
LGTM, modulo one nit

File: cpp/src/arrow/util/compression_lz4.cc
Comment (bkietz) [reply cpp/src/arrow/util/compression_lz4.cc:338]: > I'm not able to find kDefaultCompressionLevel, also not in the code for other compressors.

First of all, sorry for the confusion: Instead of `kDefaultCompressionLevel` I should have written `kLz4DefaultCompressionLevel`. My intent was to indicate that all of these should result in an equivalent codec:
1. `Lz4Codec()`
2. `Lz4Codec(kUseDefaultCompressionLevel)`
3. `Lz4Codec(kLz4DefaultCompressionLevel)`

However as you have it written `1` results in an Lz4Codec with `compression_level_ == kUseDefaultCompressionLevel` whereas `2,3` result in `compression_level_ == kLz4DefaultCompressionLevel`.

---

My stylistic preference would be to delete the default constructor of `Lz4Codec` and write `Lz4HadoopCodec`'s constructor to explicitly pass `kUseDefaultCompressionLevel` to the base class' constructor like so:

```c++
Lz4HadoopCodec() : Lz4Codec(kUseDefaultCompressionLevel) {}
```

**Comment by bkietz** [comment]
@joosthooz looks like you've still got a lint failure and there's one more constant to sometimes hardcode for older liblz4 https://github.com/apache/arrow/runs/4436211338?check_suite_focus=true#step:8:1341

**Comment by bkietz** [comment]
@github-actions autotune everything

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by pitrou — APPROVED** [review]
+1, thanks a lot for the update @joosthooz . Will wait for CI and merge if green.

**Comment by pitrou** [comment]
@github-actions autotune everything

**Pull request closed without merging**
